### ISSUE TITLE
Now getting FSDP strategy when requesting HSDP strategy for Llama3 models due to changes in  parallelize_llama.py

### ISSUE THREAD

**Issue by walkup** [pytorch/torchtitan#711]
With the current (Dec. 2, 2024) torchtitan, attempts to specify HSDP for llama3_8b are actually resulting in a fully-sharded strategy.  When using the llama3_8b.toml training config : 

[training]
batch_size = 2
seq_len = 8192
warmup_steps = 200  # lr scheduler warm up
max_norm = 1.0  # grad norm clipping
steps = 1000
data_parallel_replicate_degree = 2
data_parallel_shard_degree = 8
tensor_parallel_degree = 1
compile = true
dataset = "c4"

I expect a two-node test to use parameter sharding within each node; and that was observed with a Sept. 18 version of torchtitan.  However, with the current version, there is only one communicator formed with all 16 ranks, and model parameters appear to be sharded across all 16 ranks, ignoring the specified replicate degree.  The different behavior is very apparent in pytorch profiles.  The issue can be traced to changes in parallelize_llama.py.  If I replace the code section before apply_fsdp(...) in the current parallelize_llama.py with the simpler code from the Sept. 18 version : 

    if parallel_dims.dp_shard_enabled:
        if parallel_dims.dp_replicate_enabled:
            dp_mesh = world_mesh["dp_replicate", "dp_shard"]
        else:
            dp_mesh = world_mesh["dp"]

        apply_fsdp(...)

hybrid sharding works as expected, with parameter sharding within each node, and calls to NCCL AllReduce to sum gradients across nodes.  The current torchtitan code apparently supports additional features such as a context parallel method, but I would still expect to get  hybrid strategy when the training section of the config file is as listed above.  

Are changes to the training config file needed to recover hybrid sharding ... or is there really an issue with parallelize_llama.py that is currently resulting in FSDP strategy when the user specifies a replicate degree > 1 ?

**Comment by XilunWu** [comment]
I think the issue is likely introduced in https://github.com/pytorch/torchtitan/pull/683 . Really appreciate the timely reporting!
